### PR TITLE
Adds pre-validation before issuing an RPC so that the sender knows they have issued an RPC from an incorrect role

### PR DESCRIPTION
**Pull request by kberg-amzn**
Signed-off-by: kberg-amzn <karlberg@amazon.com>

This adds a pre-validation check on the invocation side to ensure that if a user works around our guard rails and issues an RPC from an incorrect role, they get immediate feedback and output in the console log so they are aware they have made a mistake. This supplements the more rigorous security checks made my the EntityReplicator inside EntityReplicator::ValidateRpcMessage(), which will drop the RPC on the receiver, and potentially even disconnect, but otherwise won't provide any feedback to the invoker that a mistake has been made.

### CODE REVIEW

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja [hunk 1]
```diff
@@ -324,14 +324,31 @@ void {{ ClassName }}::Set{{ UpperFirst(Property.attrib['Name']) }}(const {{ Prop
 {{     AutoComponentMacros.ParseRpcParams(Property, paramNames, paramTypes, paramDefines) }}
 void {{ ClassName }}::{{ UpperFirst(Property.attrib['Name']) }}({{ PrintRpcParameters('', paramDefines) }})
 {
+{% if (InvokeFrom == 'Authority') %}
+{#  InvokeFrom Authority inserts the invocation into the controller, not the component #}
+    if (!IsAuthority())
+{% elif (InvokeFrom == 'Autonomous') %}
+{#  InvokeFrom Authority inserts the invocation into the controller, not the component #}
+    if (!IsAutonomous())
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja:332]: This really hammers home that these should be updated to IsNetEntityRoleAuthority/Autonomous

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja [hunk 2]
```diff
@@ -324,14 +324,31 @@ void {{ ClassName }}::Set{{ UpperFirst(Property.attrib['Name']) }}(const {{ Prop
...
+{% elif (InvokeFrom == 'Server') %}
+    if (!IsNetEntityRoleServer())
+{% elif (InvokeFrom == 'Client') %}
+    if (!IsNetEntityRoleClient())
+{% endif %}
+    {
+        const char* currentRoleString = GetNetBindComponent() ? GetEnumString(GetNetBindComponent()->GetNetEntityRole()) : "NetBindComponent is nullptr";
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja:339]:
```suggestion
        const char* currentRoleString = GetNetBindComponent() ? GetEnumString(GetNetBindComponent()->GetNetEntityRole()) : "an undetermined role, NetBindComponent is nullptr";
```

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja (hunk 1 above)
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja:332]: Yup, consistency would be nice. Luckily Gene seems to have opened a pull-request for this just a few seconds ago 🤣

File: Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja (hunk 2 above)
Comment (kberg-amzn) [reply Gems/Multiplayer/Code/Include/Multiplayer/AutoGen/AutoComponent_Source.jinja]: 👍

**Review by AMZN-puvvadar — APPROVED** [review]

**Review by AMZN-Gene — APPROVED** [review]

**Pull request merged**